### PR TITLE
Enable Fields selection for autogenerated Reads

### PR DESCRIPTION
**Pull request by Calebjh**
Breaking Change to the contract for the DefaultRead{Type} generated functions. Now has a `preload bool` argument which determines whether or not to enable AUTO_PRELOAD within the read. Then the autogenerated service checks if there is a `{infoblox.api.}FieldSelection` field in the Read{Type}Request and generates proper handling if there is, otherwise generating nothing special.

I'd like to see this propagated to the List behavior eventually, too (executing the collection operator specific code in the autogenerated service instead of DefaultList{Type}) since this allows for that code to only be generated if collection operators are present in the List request, and offers better separation between vanilla protoc-gen-gorm code and code that ties into the toolkit.

### CODE REVIEW

File: plugin/servergen.go [hunk 1]
```diff
@@ -361,3 +372,16 @@ func (p *OrmPlugin) generatePreserviceHook(svc, typeName, inTypeName, mthd strin
 	p.P(`Before`, mthd, `(context.Context, *`, inTypeName, `, *`, p.Import(gormImport), `.DB) (context.Context, *`, p.Import(gormImport), `.DB, error)`)
 	p.P(`}`)
 }
+
+func (p *OrmPlugin) hasFieldsSelecter(object generator.Object) string {
```
Comment (Evgeniy-L) [inline plugin/servergen.go]: Is spelling correct? Shouldn't it be `hasFieldsSelector`?

**Comment by askurydzin** [comment]
what will happen if we set auto_preload twice? if nothing, we can avoid introducing breaking changes (however it should not hurt since maj version is < 1).

**Comment by dkukharau** [comment]
I'm wondering if the original issue is critical enough so we can merge this unwanted solution instead of preparing the proper one? Doesn't look so for me.

**Comment by Calebjh** [comment]
So, I did some refactoring for twofold reasons, 
1) so that I could preload the services/methods and associate them with the Ormable objects
2) to bring it closer to a unit-testable state

The refactoring was purely under-the-hood, generating the same code, and then I added an additional `DefaultRead{Object}Fields(..., fs &query.FieldSelection)` when a `FieldSelection` object is present in the read RPC for the object (it will only track the last rpc defined for each verb, although this maybe should be extended). This will be called in the autogenerated Read RPC functions. Everywhere else can/does use the unchanged `DefaultRead{Object}` to autopreload.

I would consider this **no longer a breaking change**.

File: plugin/servergen.go [hunk 2]
```diff
@@ -7,52 +7,130 @@ import (
...
-func (p *OrmPlugin) generateDefaultServer(file *generator.FileDescriptor) {
+const (
+	createService = "Create"
+	readService   = "Read"
+	updateService = "Update"
+	deleteService = "Delete"
+	listService   = "List"
+)
+
+var ormableServices []autogenService
```
[the remark this answers is not in the document]
Comment (Calebjh) [reply plugin/servergen.go]: Ah, to be honest I had forgotten that the ormableTypes was (or had been made) a field of the plugin as opposed to being a global and thought it was consistent with that.

**Comment by Calebjh** [comment]
It was both of those reasons, hide collection operators from those not using them (by generating the second read function only if they're being used) and maintain backward compatibility. I actually was unifying it into a single Read function (with the fields object as an argument depending on the usage of the collection operator), but we've been making a lot of breaking changes recently, so I thought this might be safer... it would be pretty easy to unify them though, as that was a major point of the refactoring.

**Comment by dkukharau** [comment]
Then let's do it! If one doesn't use field selection - he will not be broken, if uses - then he would love these changes.

**Pull request merged**